### PR TITLE
Add a queryFilter function to Daml.Script

### PR DESCRIPTION
**Pull request by bame-da**
CHANGELOG_BEGIN
 - Add `queryFor` to the Script standard library. Combines `query` and filter. A very common pattern.
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — COMMENTED** [review]
Does this really cross the fairbairn threshold? Happy to be convinced it does but there is an endless number of combinators you can construct by applying `filter` and `map` to the result of something else.

**Comment by bame-da** [comment]
I would argue that yes it does. I can't think of many circumstances where I'd in practice use `query` without following it immediately with a `filter`. Ie I think `queryFor` is the more common function than `query`.

**Comment by cocreature** [comment]
Sounds reasonable, I still feel a bit uneasy about the name. `for` as a suffix for filtering isn’t used on any of our ledger APIs (including the JSON API). Nor is it used in `daml-stdlib`. Unfortunately I don’t have a great suggestion for an alternative, a few options:

1. Somehow try to overload `query`. This matches the behavior of the JSON API but all options I can come up with lead to a huge mess and a DAML API that is far from idiomatic.
2. `queryFilter` or `filterQuery`. I don’t think we have either naming scheme in `daml-stdlib` atm but it’s fairly obvious what it does.
3. `queryOptional`. This would be different from what you implemented but it matches `mapOptional` and is more powerful in that you can change the result type without resorting to something unsafe like `fromSome`. You probably still want something like the `queryFor` you implemented on top of this so it doesn’t really solve the issue.
4. `queryFor` as used in this PR: Maybe users can guess what it does but it’s very different from `forA` which seems potentially confusing.

I have a slight preference for `queryFilter` as the one that seems the most consistent with other naming schemes but I could definitely be convinced of something else.

**Comment by bame-da** [comment]
`queryFilter` works for me. I didn't put much thought into `queryFor`, but decided against `by` or `on` as it's not a generalization of `query`.

**Comment by hurryabit** [comment]
@bame-da `queryFor` is a generalization of `query`:
```haskell
query p == queryFor p (\_ -> True)
```
Thus, `queryBy` might be an ok name.

**Comment by cocreature** [comment]
I don’t think `By` makes sense as a suffix. We use `By` to provide a custom implemenation for things that are otherwise specified based on the type and the corresponding typeclass instance. That’s not what is happening here.

**Comment by bame-da** [comment]
Yes, sorry, I wasn't being specific enough. I meant `queryFor` is not a generalization of `query` in the same way `sortBy` is a generalization of `sort`. @cocreature put my vague thoughts into more precise words.

**Comment by cocreature** [comment]
Based on the discussion, I would suggest to rename this to `queryFilter` and then I’m fine with merging it.

File: daml-script/daml/Daml/Script.daml [hunk 1]
```diff
@@ -108,6 +108,11 @@ data QueryACS a = QueryACS
 query : forall t. Template t => Party -> Script [(ContractId t, t)]
 query p = lift $ Free $ Query (QueryACS p (templateTypeRep @t) (pure . map (\(cid, tpl) -> (coerceContractId cid, fromSome $ fromAnyTemplate tpl))))
 
+-- | Query the set of active contracts of the template
+-- that are visible to the given party and match the given predicate.
+queryFor : Template c => Party -> (c -> Bool) -> Script [(ContractId c, c)]
+queryFor p f = filter (\(_, c) -> f c) <$> query p
```
Comment (bame-da) [inline daml-script/daml/Daml/Script.daml:114]:
```suggestion
queryFilter : Template c => Party -> (c -> Bool) -> Script [(ContractId c, c)]
queryFilter p f = filter (\(_, c) -> f c) <$> query p
```

**Review by cocreature — APPROVED** [review]
Thanks!

**Pull request merged**